File: app/services/auth_service.py
import httpx
from datetime import datetime, timedelta, timezone
from jose import JWTError, jwt
from passlib.context import CryptContext
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import HTTPException, status

from app.config import SECRET_KEY, ALGORITHM, ACCESS_TOKEN_EXPIRE_MINUTES, GOOGLE_CLIENT_ID, GITHUB_CLIENT_ID, GITHUB_CLIENT_SECRET
from app.models.user import User

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def hash_password(password: str) -> str:
    return pwd_context.hash(password)


def verify_password(plain: str, hashed: str) -> bool:
    return pwd_context.verify(plain, hashed)


def create_access_token(data: dict) -> str:
    to_encode = data.copy()
    expire = datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire})
    return jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)


def decode_access_token(token: str) -> dict:
    try:
        return jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
    except JWTError:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token")


async def register_user(db: AsyncSession, email: str, password: str, full_name: str | None = None) -> User:
    result = await db.execute(select(User).where(User.email == email))
    if result.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Email already registered")

    user = User(
        email=email,
        hashed_password=hash_password(password),
        full_name=full_name,
        auth_provider="email",
    )
    db.add(user)
    await db.commit()
    await db.refresh(user)
    return user


async def authenticate_user(db: AsyncSession, email: str, password: str) -> User:
    result = await db.execute(select(User).where(User.email == email))
    user = result.scalar_one_or_none()
    if not user or not user.hashed_password or not verify_password(password, user.hashed_password):
        raise HTTPException(status_code=401, detail="Invalid email or password")
    return user


async def _get_google_public_keys() -> dict:
    async with httpx.AsyncClient() as client:
        resp = await client.get("https://www.googleapis.com/oauth2/v3/certs")
        resp.raise_for_status()
        return resp.json()


async def _exchange_github_code(code: str, redirect_uri: str | None = None) -> str:
    if not GITHUB_CLIENT_ID or not GITHUB_CLIENT_SECRET:
        raise HTTPException(status_code=500, detail="GitHub OAuth is not configured")

    payload = {
        "client_id": GITHUB_CLIENT_ID,
        "client_secret": GITHUB_CLIENT_SECRET,
        "code": code,
    }
    if redirect_uri:
        payload["redirect_uri"] = redirect_uri

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            "https://github.com/login/oauth/access_token",
            headers={"Accept": "application/json"},
            data=payload,
            timeout=20,
        )
        resp.raise_for_status()
        data = resp.json()

    token = data.get("access_token")
    if not token:
        raise HTTPException(status_code=401, detail="Invalid GitHub code or token exchange failed")
    return token


async def github_login(db: AsyncSession, code: str, redirect_uri: str | None = None) -> User:
    token = await _exchange_github_code(code, redirect_uri)

    async with httpx.AsyncClient() as client:
        user_resp = await client.get(
            "https://api.github.com/user",
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/vnd.github+json",
            },
            timeout=20,
        )
        user_resp.raise_for_status()
        github_user = user_resp.json()

        email = github_user.get("email")
        if not email:
            emails_resp = await client.get(
                "https://api.github.com/user/emails",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github+json",
                },
                timeout=20,
            )
            emails_resp.raise_for_status()
            emails = emails_resp.json()
            email = next(
                (item.get("email") for item in emails if item.get("primary") and item.get("verified")),
                None,
            ) or next((item.get("email") for item in emails if item.get("verified")), None)

        if not email:
            raise HTTPException(status_code=400, detail="GitHub account email is unavailable")

        result = await db.execute(select(User).where(User.email == email))
        user = result.scalar_one_or_none()

        if not user:
            user = User(
                email=email,
                full_name=github_user.get("name") or github_user.get("login"),
                picture=github_user.get("avatar_url"),
                auth_provider="github",
                github_token=token,
            )
            db.add(user)
            await db.commit()
            await db.refresh(user)
            return user

        user.github_token = token
        if not user.auth_provider or user.auth_provider == "email":
            user.auth_provider = "github"
        await db.commit()
        await db.refresh(user)
        return user


async def google_login(db: AsyncSession, credential: str) -> User:
    try:
        jwks = await _get_google_public_keys()
        header = jwt.get_unverified_header(credential)
        kid = header.get("kid")
        key = next((k for k in jwks["keys"] if k["kid"] == kid), None)
        if not key:
            raise ValueError("Google signing key not found")
        idinfo = jwt.decode(
            credential,
            key,
            algorithms=["RS256"],
            audience=GOOGLE_CLIENT_ID,
            issuer=["https://accounts.google.com", "accounts.google.com"],
        )
    except Exception as e:
        raise HTTPException(status_code=401, detail=f"Invalid Google token: {e}")

    email = idinfo["email"]
    result = await db.execute(select(User).where(User.email == email))
    user = result.scalar_one_or_none()

    if not user:
        user = User(
            email=email,
            full_name=idinfo.get("name"),
            picture=idinfo.get("picture"),
            auth_provider="google",
        )
        db.add(user)
        await db.commit()
        await db.refresh(user)

    return user


async def get_current_user(db: AsyncSession, token: str) -> User:
    payload = decode_access_token(token)
    user_id = payload.get("sub")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid token")
    result = await db.execute(select(User).where(User.id == int(user_id)))
    user = result.scalar_one_or_none()
    if not user:
        raise HTTPException(status_code=401, detail="User not found")
    return user

File: src/services/authService.ts
import jwt from "jsonwebtoken";
import bcrypt from "bcryptjs";
import { createRemoteJWKSet, jwtVerify } from "jose";
import { eq } from "drizzle-orm";
import { db } from "../db/connection.js";
import { users, type UserRow } from "../db/schema.js";
import {
  SECRET_KEY, ALGORITHM, ACCESS_TOKEN_EXPIRE_MINUTES,
  GOOGLE_CLIENT_ID, GITHUB_CLIENT_ID, GITHUB_CLIENT_SECRET,
} from "../config.js";

export class HttpError extends Error {
  constructor(public status: number, message: string) {
    super(message);
  }
}

export function hashPassword(password: string): string {
  return bcrypt.hashSync(password, 10);
}

export function verifyPassword(plain: string, hashed: string): boolean {
  return bcrypt.compareSync(plain, hashed);
}

export function createAccessToken(data: Record<string, unknown>): string {
  return jwt.sign(data, SECRET_KEY, { algorithm: ALGORITHM, expiresIn: `${ACCESS_TOKEN_EXPIRE_MINUTES}m` });
}

export function decodeAccessToken(token: string): jwt.JwtPayload {
  try {
    return jwt.verify(token, SECRET_KEY, { algorithms: [ALGORITHM] }) as jwt.JwtPayload;
  } catch {
    throw new HttpError(401, "Invalid token");
  }
}

// Port of app/models/user.py's Column(default=...) values — SQLAlchemy applies these at INSERT
// time on the Python side (client-side, not a real DB column default; confirmed live: the actual
// Postgres `users.is_superuser`/`is_active` columns have column_default = null), so every new-user
// insert here must set them explicitly too, or they land NULL instead of the intended default.
const NEW_USER_DEFAULTS = { isActive: true, isSuperuser: false, dateFormat: "YYYY-MM-DD", language: "en" };

export async function registerUser(email: string, password: string, fullName?: string): Promise<UserRow> {
  const existing = await db.select().from(users).where(eq(users.email, email)).limit(1);
  if (existing.length > 0) throw new HttpError(400, "Email already registered");

  const [user] = await db
    .insert(users)
    .values({ email, hashedPassword: hashPassword(password), fullName, authProvider: "email", ...NEW_USER_DEFAULTS })
    .returning();
  return user;
}

export async function authenticateUser(email: string, password: string): Promise<UserRow> {
  const [user] = await db.select().from(users).where(eq(users.email, email)).limit(1);
  if (!user || !user.hashedPassword || !verifyPassword(password, user.hashedPassword)) {
    throw new HttpError(401, "Invalid email or password");
  }
  return user;
}

async function exchangeGithubCode(code: string, redirectUri?: string): Promise<string> {
  if (!GITHUB_CLIENT_ID || !GITHUB_CLIENT_SECRET) {
    throw new HttpError(500, "GitHub OAuth is not configured");
  }
  const payload: Record<string, string> = { client_id: GITHUB_CLIENT_ID, client_secret: GITHUB_CLIENT_SECRET, code };
  if (redirectUri) payload.redirect_uri = redirectUri;

  const resp = await fetch("https://github.com/login/oauth/access_token", {
    method: "POST",
    headers: { Accept: "application/json", "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!resp.ok) throw new HttpError(401, "Invalid GitHub code or token exchange failed");
  const data = (await resp.json()) as { access_token?: string };
  if (!data.access_token) throw new HttpError(401, "Invalid GitHub code or token exchange failed");
  return data.access_token;
}

export async function githubLogin(code: string, redirectUri?: string): Promise<UserRow> {
  const token = await exchangeGithubCode(code, redirectUri);

  const userResp = await fetch("https://api.github.com/user", {
    headers: { Authorization: `Bearer ${token}`, Accept: "application/vnd.github+json" },
  });
  if (!userResp.ok) throw new HttpError(401, "Failed to fetch GitHub profile");
  const githubUser = (await userResp.json()) as { email?: string; name?: string; login?: string; avatar_url?: string };

  let email = githubUser.email;
  if (!email) {
    const emailsResp = await fetch("https://api.github.com/user/emails", {
      headers: { Authorization: `Bearer ${token}`, Accept: "application/vnd.github+json" },
    });
    if (emailsResp.ok) {
      const emails = (await emailsResp.json()) as { email: string; primary: boolean; verified: boolean }[];
      email =
        emails.find((e) => e.primary && e.verified)?.email ??
        emails.find((e) => e.verified)?.email;
    }
  }
  if (!email) throw new HttpError(400, "GitHub account email is unavailable");

  const [existing] = await db.select().from(users).where(eq(users.email, email)).limit(1);
  if (!existing) {
    const [user] = await db
      .insert(users)
      .values({
        email,
        fullName: githubUser.name || githubUser.login,
        picture: githubUser.avatar_url,
        authProvider: "github",
        githubToken: token,
        ...NEW_USER_DEFAULTS,
      })
      .returning();
    return user;
  }

  const [updated] = await db
    .update(users)
    .set({ githubToken: token, authProvider: !existing.authProvider || existing.authProvider === "email" ? "github" : existing.authProvider })
    .where(eq(users.id, existing.id))
    .returning();
  return updated;
}

const googleJwks = GOOGLE_CLIENT_ID ? createRemoteJWKSet(new URL("https://www.googleapis.com/oauth2/v3/certs")) : null;

export async function googleLogin(credential: string): Promise<UserRow> {
  if (!googleJwks) throw new HttpError(500, "Google OAuth is not configured");
  let email: string, name: string | undefined, picture: string | undefined;
  try {
    const { payload } = await jwtVerify(credential, googleJwks, {
      audience: GOOGLE_CLIENT_ID,
      issuer: ["https://accounts.google.com", "accounts.google.com"],
    });
    email = payload.email as string;
    name = payload.name as string | undefined;
    picture = payload.picture as string | undefined;
  } catch (e) {
    throw new HttpError(401, `Invalid Google token: ${e instanceof Error ? e.message : e}`);
  }

  const [existing] = await db.select().from(users).where(eq(users.email, email)).limit(1);
  if (existing) return existing;

  const [user] = await db
    .insert(users)
    .values({ email, fullName: name, picture, authProvider: "google", ...NEW_USER_DEFAULTS })
    .returning();
  return user;
}

export async function getCurrentUser(token: string): Promise<UserRow> {
  const payload = decodeAccessToken(token);
  const userId = payload.sub;
  if (!userId) throw new HttpError(401, "Invalid token");
  const [user] = await db.select().from(users).where(eq(users.id, Number(userId))).limit(1);
  if (!user) throw new HttpError(401, "User not found");
  return user;
}
